src/io/xmap.rs

```diff
@@ -0,0 +1,1357 @@
+        loop {
+            buffer.clear();
+            if stream.read_line(&mut buffer)? == 0 {
+                break;
+            };
+            if buffer.starts_with('#') {
+                header.push(String::from(buffer.trim_end()));
+            } else {
+                break;
+            }
+        }
+        Ok(Reader {
+            stream,
+            header,
+            buffer,
+        })
+    }
+
+    /// Constructs [Container](struct.Container.html) over XMAP content while consuming reader.
+    ///
+    /// # Example
+    /// ```rust
+    /// use bio::io::xmap::Reader;
+    ///
+    /// let path = "tests/resources/valid_input.xmap";
+    /// let reader = Reader::from_path(&path).unwrap();
+    /// let container = reader.into_container();
+    /// ```
+    pub fn into_container(self) -> Result<Container> {
+        let header = self.header.clone();
+        let mut inner = HashMap::new();
+        for rec in self {
+            let mut rec = match rec {
+                Ok(rec) => rec,
+                Err(e) => return Err(e),
+            };
+            inner
+                .entry(rec.contig())
+                .or_insert_with(BTreeMap::new)
+                .insert(rec.start() as u64, rec);
+        }
+        Ok(Container::new(header, inner))
+    }
+}
+
+impl<R: BufRead> XmapRead for Reader<R> {
+    fn read_into(&mut self, record: &mut Record) -> Result<bool> {
+        if self.buffer.is_empty() {
+            return Ok(false);
+        }
+        let res = match record.fill(self.buffer.trim()) {
+            Ok(()) => Ok(true),
+            Err(e) => Err(e),
+        };
+        self.buffer.clear();
+        match self.stream.read_line(&mut self.buffer) {
+            Ok(_) => res,
+            Err(e) => Err(e.into()),
+        }
+    }
+}
+
+/// Iterator over records.
+impl<R: BufRead> Iterator for Reader<R> {
+    type Item = Result<Record>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut rec = Record::new();
+        match self.read_into(&mut rec) {
+            Ok(true) => Some(Ok(rec)),
+            Ok(false) => None,
+            Err(e) => Some(Err(e)),
+        }
+    }
+}
+
+/// A container with CMAP header and records stored in B-trees.
+///
+/// Each tree corresponds to one reference contig and is ordered by start position.
+#[derive(Debug, new, PartialEq)]
+pub struct Container {
+    header: Vec<String>,
+    pos_trees: HashMap<u32, BTreeMap<u64, Record>>,
+    //     pos_trees: HashMap<u32, BTreeMap<u32, Rc<Record>>>,
+}
+
+impl Container {
+    /// Constructs a new `Container` from path.
+    pub fn from_path<P: AsRef<Path>>(path: P) -> Result<Self> {
+        let reader = Reader::from_path(path)?;
+        reader.into_container()
+    }
+
+    /// Returns the subset of records contained in given region in a contig.
+    pub fn subset_filter_by_region(
+        &mut self,
+        region_contig: u32,
+        region_start: u64,
+        region_end: u64,
+    ) -> Result<Self> {
```

Comment:
This is all we need:

```suggestion
    pub fn fetch(
        &mut self,
        region: bio_types::genome::Interval
    ) -> Result<impl Iterator<Item=Record>> {
```

Reply:
Why do we want the filter to return an iterator over `Record` elements? Why not a subtree that we can also iterate through?
Should we then merge the subset and range filter?

Reply:
We just need this fetch. Because all we need in varlociraptor is the records that belong to a particular region. Regarding the return: returning a subtree would require new memory allocations (slow). Hence, it is idiomatic to return an iterator instead, so that the caller can decide. The iterator should be easy to build here using standard btree queries.

Reply:
You're right; that sounds indeed more reasonable. 
Should I differentiate between whole contigs and particular regions or just combine the two functions?

Reply:
Replaced all 4 functions with one generic `fetch()`. For whole contigs, one can set the region start and end accordingly.